logstash-core/src/main/java/org/logstash/secret/store/backend/JavaKeyStore.java

```diff
@@ -0,0 +1,227 @@
+            bytes[i] = '\0';
+        }
+        return chars;
+    }
+
+    /**
+     * Converts characters from ascii encoded text to a byte[] and zero outs the original char[]
+     *
+     * @param chars the chars from an ascii encoded text (note - no validation is done to ensure ascii encoding)
+     * @return the corresponding byte[]
+     */
+    private byte[] asciiCharToBytes(char[] chars) {
+        byte[] bytes = new byte[chars.length];
+        for (int i = 0; i < chars.length; i++) {
+            bytes[i] = (byte) chars[i];
+            chars[i] = '\0';
+        }
+        return bytes;
+    }
+
+    /**
+     * Attempt to keep the secret data out of the heap.
+     *
+     * @param secret the secret to zero out
+     */
+    private void clearSecret(byte[] secret) {
+        if (secret != null) {
+            for (int i = 0; i < secret.length; ++i) {
+                secret[i] = '\0';
+            }
+            secret = null;
+        }
+    }
+
+    @Override
+    public Collection<SecretIdentifier> list() {
+        Set<SecretIdentifier> identifiers = new HashSet<>();
+        try {
+            readLock.lock();
+            Enumeration<String> aliases = keyStore.aliases();
+            while (aliases.hasMoreElements()) {
+                String alias = aliases.nextElement();
+                identifiers.add(SecretIdentifier.fromExternalForm(alias));
+            }
+        } catch (KeyStoreException e) {
+            e.printStackTrace();
+        } finally {
+            readLock.unlock();
+        }
+        return identifiers;
+    }
+
+    @Override
+    public void persistSecret(SecretIdentifier identifier, byte[] secret) {
+        try {
+            writeLock.lock();
+            SecretKeyFactory factory = SecretKeyFactory.getInstance("PBE");
+            //PBEKey requires an ascii password, so base64 encode it
+            byte[] base64 = Base64.getEncoder().encode(secret);
+            PBEKeySpec passwordBasedKeySpec = new PBEKeySpec(asciiBytesToChar(base64));
+            SecretKey secretKey = factory.generateSecret(passwordBasedKeySpec);
+            keyStore.setEntry(identifier.toExternalForm(), new KeyStore.SecretKeyEntry(secretKey), protectionParameter);
+            try (final OutputStream os = Files.newOutputStream(keyStorePath)) {
+                keyStore.store(os, keyStorePass);
+            } finally {
+                passwordBasedKeySpec.clearPassword();
+                clearSecret(secret);
+            }
+        } catch (Exception e) {
+            throw new SecretStoreException.PersistException(identifier, e);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void purgeSecret(SecretIdentifier identifier) {
+        try {
+            writeLock.lock();
+            try (final InputStream is = Files.newInputStream(keyStorePath)) {
+                keyStore.load(is, keyStorePass);
+                keyStore.deleteEntry(identifier.toExternalForm());
```

Comment:
Let's add a logger.trace call here to note what is deleted.

Reply:
yeah, i need some more debug logging. Also noticed some e.printStackTrace() snuck in here... will beef up logging and error handling a bit more.